File: algorithm/baekjoon/src/main/java/com/keunwon/algorithm/dfs/Problem2606.java
package com.keunwon.algorithm.dfs;

import java.util.Scanner;

public class Problem2606 {
    private static int V;
    private static int E;
    private static int[][] map;
    private static boolean[] checked;
    private static int count = 0;

    public static void main(String[] args) {
        input();


        for (int i = 1; i < V + 1; i++) {
            if (map[1][i] == 1 && !checked[i]) {
                dfs(map[1][i]);
            }
        }

        System.out.println(count);
    }

    private static void dfs(int x) {
        checked[x] = true;

        for (int i = 1; i < V + 1; i++) {
            if (map[x][i] == 1 && !checked[i]) {
                count++;
                dfs(i);
            }
        }
    }

    private static void input() {
        Scanner sc = new Scanner(System.in);

        V = sc.nextInt();
        E = sc.nextInt();
        map = new int[V + 1][V + 1];
        checked = new boolean[V + 1];

        for (int i = 0; i < E; i++) {
            int v1 = sc.nextInt();
            int v2 = sc.nextInt();

            map[v1][v2] = map[v2][v1] = 1;
        }
    }
}

File: algorithm/kotlin-algorithm/src/main/kotlin/com/keunwon/algorithm/baekjoon/Problem2606.kt
package com.keunwon.algorithm.baekjoon

import java.util.*

/**
 * Title: 바이러스
 * Level: 실버-3
 **/
class Problem2606 {
    private lateinit var graph: Array<MutableList<Int>>

    fun solution(n: Int, arr: Array<Pair<Int, Int>>): Int {
        this.graph = Array(n + 1) { mutableListOf<Int>() }.apply {
            arr.forEach { (a, b) ->
                this[a].add(b)
                this[b].add(a)
            }
        }
        val visited = BooleanArray(n + 1).apply { this[1] = true }
        val queue = LinkedList<Int>().apply { offer(1) }

        while (queue.isNotEmpty()) {
            val cur = queue.poll()

            for (next in graph[cur]) {
                if (visited[next]) continue

                visited[next] = true
                queue.offer(next)
            }
        }
        return visited.count { it } - 1
    }
}

fun main() = System.`in`.bufferedReader().use { br ->
    System.out.bufferedWriter().use { bw ->
        val n = br.readLine().toInt()
        val m = br.readLine().toInt()
        val arr = Array(m) {
            br.readLine()
                .split(" ")
                .let { it[0].toInt() to it[1].toInt() }
        }

        Problem2606().solution(n, arr).also { bw.write("$it") }
        bw.flush()
    }
}
